### PR TITLE
fix(NODE-3815): update Decimal128 constructor validation

### PR DESCRIPTION
**Pull request by syz99**
#### What is changing?
Add new checkings/exceptions for invalid types

##### Is there new documentation needed for these changes?
n/a

### Double check the following

- [x] Ran `npm run lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/node/decimal128_tests.js [hunk 1]
```diff
@@ -1216,6 +1217,24 @@ describe('Decimal128', function () {
...
+
+  it('throws appropriate error for invalid constructor arguments', () => {
+    const byteLengthErrMsg = 'Decimal128 must take a Buffer of 16 bytes';
+    const constructorArgErrMsg = 'Decimal128 must take a Buffer or string';
+
+    expect(() => new Decimal128(-0)).to.throw(constructorArgErrMsg);
+    expect(() => new Decimal128(-1)).to.throw(constructorArgErrMsg);
+    expect(() => new Decimal128(10)).to.throw(constructorArgErrMsg);
+    expect(() => new Decimal128(1111111111111111)).to.throw(constructorArgErrMsg);
+
```
Comment (baileympearson) [inline test/node/decimal128_tests.js:1230]: Could we break this test into two parts - one for the tests that test the constructor argument error message and one for the byteLengthErrorMessage?

File: src/bson.ts [hunk 2]
```diff
@@ -112,7 +112,7 @@ export interface Document {
 const MAXSIZE = 1024 * 1024 * 17;
 
 // Current Internal Temporary Serialization Buffer
-let buffer = Buffer.alloc(MAXSIZE);
+let buffer = Buffer.allocUnsafe(MAXSIZE);
```
Comment (baileympearson) [inline src/bson.ts:115]: Is this change intentional?  It doesn't seem to be related to adding new type errors to the Decimal128 constructor

**Review by baileympearson — CHANGES_REQUESTED** [review]

**Review by nbbeeken — DISMISSED** [review]
lgtm 👍

File: test/node/decimal128_tests.js [hunk 3]
```diff
@@ -1216,6 +1216,27 @@ describe('Decimal128', function () {
...
+  it('throws correct error for an invalid Buffer constructor argument', () => {
+    const byteLengthErrMsg = 'Decimal128 must take a Buffer of 16 bytes';
+
+    expect(() => new Decimal128(new Uint8Array(0))).to.throw(byteLengthErrMsg);
+    expect(() => new Decimal128(Buffer.alloc(0))).to.throw(byteLengthErrMsg);
+    expect(() => new Decimal128(new Uint8Array(3))).to.throw(byteLengthErrMsg);
+    expect(() => new Decimal128(Buffer.alloc(3))).to.throw(byteLengthErrMsg);
+    expect(() => new Decimal128(new Uint8Array(17))).to.throw(byteLengthErrMsg);
+    expect(() => new Decimal128(Buffer.alloc(17))).to.throw(byteLengthErrMsg);
+    new Decimal128(Buffer.alloc(16));
```
Comment (durran) [inline test/node/decimal128_tests.js:1239]: Are these here because we don't expect errors? If so then I'd probably prefer explicitly asserting in a separate `it` block for example:

```js
it('does not throws error for an empty Buffer constructor argument of correct length', () => {
  expect(() => new Decimal128(Buffer.alloc(16))).to.not.throw();
  expect(() => new Decimal128(new Uint8Array(16))).to.not.throw();
});
```

**Review by durran — CHANGES_REQUESTED** [review]

**Review by durran — APPROVED** [review]

File: test/node/decimal128_tests.js (hunk 3 above)
Comment (durran) [reply test/node/decimal128_tests.js]: Thanks!

**Review by nbbeeken — APPROVED** [review]
